### ISSUE TITLE
Tests fail under zsh 5.0.7

### ISSUE THREAD

**Issue by danielshahaf** [zsh-users/zsh-syntax-highlighting#665]
Travis build fails on zsh-5.0.7 and older since commit 8b2768ab4007d3d883c3565e7998c431ba576b51, which added a test for 07f259f653df042e599975261cac7be1bda17.  (I didn't notice that yesterday because the build just before that one failed too for unrelated reasons.)

On 5.0.7 the failure is as follows:

```pod
# backslash
1..2
not ok 1 - [1,1] «\» - expected (1 1 "unknown-token"), observed (1 1 "arg0_").
ok 2 - cardinality check
```

We should fix this, for LTS distros' users and to make the CI bots useful.

**Comment by danielshahaf** [comment]
I've captured a trace with an affected zsh as of commit 2d0dddf58bab0bd9220f29919065ff85db66390e (it's appended).

tl;dr: `_zsh_highlight_main__type` is called first with the argument `\`, then that gets path-expanded and `_zsh_highlight_main__type` is called a second time with the argument `''` (empty string).  `_zsh_highlight_main__type` returns false, but the calling code doesn't check that and proceeds to set `res=''`, which in turn makes `arg0_$res` expand to `arg0_`, which is the observed result.

I think there are several different aspects here:

- [ ] 1. If the last word in the buffer is an unquoted `\` (i.e., an incomplete word), that word gets treated as though it was `''`.
- [ ] 2. `BUFFER=': release.md\'` uses `path` highlighting, and underlines the backslash too.  The backslash shouldn't be underlined.
- [ ] 3. (Actually, trailing slashes are silently ignored in general.  Does that manifest in ways more interesting than the above bullets?)
- [ ] 4. `_zsh_highlight_main__type`'s callers don't check its return status; they just assume it succeeds.
- [ ] 5. For some reason, `_zsh_highlight_main__type ""` doesn't return `none` even though `type -w ""` dutifully prints `: none`.  Probably due to unintended elision?

```zsh
+_zsh_highlight_main_highlighter_highlight_list:160> _zsh_highlight_main__type '\'
+_zsh_highlight_main__type:1> integer -r 'aliases_allowed=1'
+_zsh_highlight_main__type:7> integer 'may_cache=1'
+_zsh_highlight_main__type:10> ((  1  ))
+_zsh_highlight_main__type:11> REPLY=''
+_zsh_highlight_main__type:12> [[ -n '' ]]
+_zsh_highlight_main__type:18> ((  0  ))
+_zsh_highlight_main__type:21> unset REPLY
+_zsh_highlight_main__type:22> zmodload -e zsh/parameter
+_zsh_highlight_main__type:23> ((  0  ))
+_zsh_highlight_main__type:26> ((  0  ))
+_zsh_highlight_main__type:28> [[ '\' == *.* ]]
+_zsh_highlight_main__type:30> ((  0  ))
+_zsh_highlight_main__type:32> ((  0  ))
+_zsh_highlight_main__type:34> ((  0  ))
+_zsh_highlight_main__type:36> ((  0  ))
+_zsh_highlight_main__type:48> [[ '\' != */* ]]
+_zsh_highlight_main__type:52> REPLY=none
+_zsh_highlight_main__type:55> ((  1  ))
+_zsh_highlight_main__type:73> ((  may_cache  ))
+_zsh_highlight_main__type:73> ((  1  ))
+_zsh_highlight_main__type:74> _zsh_highlight_main__command_type_cache[(e)$1]=none
+_zsh_highlight_main__type:76> [[ -n none ]]
+_zsh_highlight_main__type:77> return 0
+_zsh_highlight_main_highlighter_highlight_list:161> local 'res=none'
+_zsh_highlight_main_highlighter_highlight_list:162> [[ none == alias ]]
+_zsh_highlight_main_highlighter_highlight_list:193> _zsh_highlight_main_highlighter_expand_path '\'
+_zsh_highlight_main_highlighter_expand_path:2> ((  1 == 1  ))
+_zsh_highlight_main_highlighter_expand_path:5> setopt localoptions nonomatch
+_zsh_highlight_main_highlighter_expand_path:6> unset REPLY
+_zsh_highlight_main_highlighter_expand_path:7> :
+_zsh_highlight_main_highlighter_highlight_list:194> _zsh_highlight_main__type '' 0
+_zsh_highlight_main__type:1> integer -r 'aliases_allowed=0'
+_zsh_highlight_main__type:7> integer 'may_cache=1'
+_zsh_highlight_main__type:10> ((  1  ))
+_zsh_highlight_main__type:11> REPLY=''
+_zsh_highlight_main__type:12> [[ -n '' ]]
+_zsh_highlight_main__type:18> ((  0  ))
+_zsh_highlight_main__type:21> unset REPLY
+_zsh_highlight_main__type:22> zmodload -e zsh/parameter
+_zsh_highlight_main__type:23> ((  0  ))
+_zsh_highlight_main__type:26> ((  0  ))
+_zsh_highlight_main__type:28> [[ '' == *.* ]]
+_zsh_highlight_main__type:30> ((  0  ))
+_zsh_highlight_main__type:32> ((  0  ))
+_zsh_highlight_main__type:34> ((  0  ))
+_zsh_highlight_main__type:36> ((  0  ))
+_zsh_highlight_main__type:48> [[ '' != */* ]]
+_zsh_highlight_main__type:55> ((  0  ))
+_zsh_highlight_main__type:66> REPLY=+_zsh_highlight_main__type:66> :
+_zsh_highlight_main__type:66> ((  aliases_allowed  ))
+_zsh_highlight_main__type:66> LC_ALL=C +_zsh_highlight_main__type:66> type -w --
+_zsh_highlight_main__type:66> REPLY=''
+_zsh_highlight_main__type:67> [[ '' == alias ]]
+_zsh_highlight_main__type:73> ((  may_cache  ))
+_zsh_highlight_main__type:73> ((  1  ))
+_zsh_highlight_main__type:74> _zsh_highlight_main__command_type_cache[(e)$1]=''
+_zsh_highlight_main__type:76> [[ -n '' ]]
+_zsh_highlight_main__type:77> return 1
+_zsh_highlight_main_highlighter_highlight_list:195> res=''
+_zsh_highlight_main_highlighter_highlight_list:200> _zsh_highlight_main__is_redirection '\'
+_zsh_highlight_main__is_redirection:6> [[ '\' == (<0-9>|)(\<|\>)* ]]
+_zsh_highlight_main_highlighter_highlight_list:209> [[ '\' == {([A-Za-z_][A-Za-z0-9_]*|[0-9]+)} ]]
+_zsh_highlight_main_highlighter_highlight_list:221> '(anon)'
+(anon):5> local -a match mbegin mend
+(anon):6> local MATCH
+(anon):6> integer MBEGIN MEND
+(anon):7> local parameter_name
+(anon):8> [[ '\' == \$ ]]
+(anon):10> [[ '\' == \$ ]]
+(anon):13> [[ '' == none ]]
+_zsh_highlight_main_highlighter_highlight_list:255> ((  ! in_redirection  ))
+_zsh_highlight_main_highlighter_highlight_list:256> [[ :start::start_of_pipeline: == *:sudo_opt:* ]]
+_zsh_highlight_main_highlighter_highlight_list:289> [[ :start::start_of_pipeline: == *:sudo_arg:* ]]
+_zsh_highlight_main_highlighter_highlight_list:296> [[ -n '' ]]
+_zsh_highlight_main_highlighter_highlight_list:317> ((  in_redirection ))
+_zsh_highlight_main_highlighter_highlight_list:317> [[ :start::start_of_pipeline: == *:always:* ]]
+_zsh_highlight_main_highlighter_highlight_list:321> ((  in_redirection ))
+_zsh_highlight_main_highlighter_highlight_list:321> [[ :start::start_of_pipeline: == *:start:* ]]
+_zsh_highlight_main_highlighter_highlight_list:322> ((  0  ))
+_zsh_highlight_main_highlighter_highlight_list:330> case  (reserved)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (suffix alias)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (alias)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (builtin)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (function)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (command)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (hashed)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (none)
+_zsh_highlight_main_highlighter_highlight_list:330> case  (*)
+_zsh_highlight_main_highlighter_highlight_list:479> _zsh_highlight_main_add_region_highlight 0 1 arg0_
```

**Comment by danielshahaf** [comment]
533bfa011622f2eb829e1310db0860ae8d3d75f3 has fixed (0) and (5).  I've filed #666 for (2).  That leaves (1) (3) (4).

**Comment by danielshahaf** [comment]
@phy1729 Thoughts on cases (1) and (4) in https://github.com/zsh-users/zsh-syntax-highlighting/issues/665#issuecomment-573435549?  Are there any improvements that can be made, or shall we close this issue?
